### PR TITLE
Added additional methods to Mark-,Pupil-,Clazz-Service. Added necessa…

### PR DESCRIPTION
**Pull request by zasamen**
…ry specifications. Issued complex criteria

### CODE REVIEW

File: server/src/main/java/school/journal/controller/MarkAPIController.java [hunk 1]
```diff
@@ -119,4 +121,80 @@ public ResponseEntity getOne(HttpServletRequest request,
...
+        }
+        return resultResponse;
+    }
+
+    @RequestMapping(method = RequestMethod.GET)
+    @ResponseBody
+    public ResponseEntity getMarksForTermOrder(
+            HttpServletRequest request,
+            @RequestParam(value = "classId") int classId,
+            @RequestParam(value = "startTerm") Date startTerm,
```
Comment (mbarouski) [inline server/src/main/java/school/journal/controller/MarkAPIController.java]: Date -> String

File: server/src/main/java/school/journal/controller/MarkAPIController.java [hunk 2]
```diff
@@ -119,4 +121,80 @@ public ResponseEntity getOne(HttpServletRequest request,
...
+            @RequestParam(value = "endTerm") Date endTerm)
+            throws ControllerException {
+        ResponseEntity resultResponse;
+        try {
+            LOGGER.info("Get mark entities by class and time terms Controller method");
+            resultResponse = new ResponseEntity(markService.
+                    getMarksForTermOrder(classId, startTerm, endTerm),
```
Comment (mbarouski) [inline server/src/main/java/school/journal/controller/MarkAPIController.java]: I think for simplicity out app will work with one term, so del startTerm and endTerm

File: server/src/main/java/school/journal/controller/MarkAPIController.java [hunk 3]
```diff
@@ -119,4 +121,80 @@ public ResponseEntity getOne(HttpServletRequest request,
...
+                    HttpStatus.OK);
+        } catch (ServiceException exc) {
+            LOGGER.error(exc);
+            resultResponse = new ResponseEntity(
+                    new ErrorObject("Error in class getting"),
```
Comment (mbarouski) [inline server/src/main/java/school/journal/controller/MarkAPIController.java]: msg text is incorrect

File: server/src/main/java/school/journal/controller/MarkAPIController.java [hunk 4]
```diff
@@ -119,4 +121,80 @@ public ResponseEntity getOne(HttpServletRequest request,
...
+        }
+        return resultResponse;
+    }
+
+    @RequestMapping(method = RequestMethod.GET)
+    @ResponseBody
+    public ResponseEntity getMarksForPupil(
+            HttpServletRequest request,
+            @RequestParam(value = "pupilId") int pupilId,
+            @RequestParam(value = "startTerm") Date startTerm,
```
Comment (mbarouski) [inline server/src/main/java/school/journal/controller/MarkAPIController.java]: the same for term dates

File: server/src/main/java/school/journal/controller/PupilAPIController.java [hunk 5]
```diff
@@ -25,101 +29,148 @@
     @Qualifier("PupilService")
     private IPupilService pupilService;
 
-    @RequestMapping(method = RequestMethod.GET)
+    @RequestMapping(method = GET)
     @ResponseBody
     public ResponseEntity get(HttpServletRequest request)
```
Comment (mbarouski) [inline server/src/main/java/school/journal/controller/PupilAPIController.java:34]: change this method to getting of list of pupils that has no class

File: server/src/main/java/school/journal/controller/PupilAPIController.java [hunk 6]
```diff
@@ -25,101 +29,148 @@
...
         } catch (Exception exc) {
             LOGGER.error(exc);
-            resultResponse = new ResponseEntity(new ErrorObject("Some critical error"), HttpStatus.INTERNAL_SERVER_ERROR);
+            resultResponse = new ResponseEntity(new ErrorObject("Some critical error"), INTERNAL_SERVER_ERROR);
         }
         return resultResponse;
     }
 
-    @RequestMapping(value = "/{pupilId}", method = RequestMethod.DELETE)
+    @RequestMapping(value = "/{classId}", method = DELETE)
```
Comment (mbarouski) [inline server/src/main/java/school/journal/controller/PupilAPIController.java]: classId -> pupilId

File: server/src/main/java/school/journal/controller/PupilAPIController.java [hunk 7]
```diff
@@ -25,101 +29,148 @@
...
         } catch (Exception exc) {
             LOGGER.error(exc);
-            resultResponse = new ResponseEntity(new ErrorObject("Some critical error"), HttpStatus.INTERNAL_SERVER_ERROR);
+            resultResponse = new ResponseEntity(new ErrorObject("Some critical error"), INTERNAL_SERVER_ERROR);
         }
         return resultResponse;
     }
 
-    @RequestMapping(value = "/{pupilId}", method = RequestMethod.GET)
+    @RequestMapping(value = "/{classId}", method = GET)
```
Comment (mbarouski) [inline server/src/main/java/school/journal/controller/PupilAPIController.java]: classId -> pupilId

File: server/src/main/java/school/journal/controller/PupilAPIController.java [hunk 8]
```diff
@@ -25,101 +29,148 @@
...
+    }
+
+    @RequestMapping(value = "/{classId}", method = GET)
+    @ResponseBody
+    public ResponseEntity getListOfPupils(HttpServletRequest request,
+                                          @PathVariable("classId") int classId)
+            throws ControllerException {
+        ResponseEntity resultResponse;
+        try {
+            LOGGER.info("Get Pupil entities Controller method");
```
Comment (mbarouski) [inline server/src/main/java/school/journal/controller/PupilAPIController.java]: use more informative messages in `.info()` or not use it

File: server/src/main/java/school/journal/repository/specification/mark/MarkSpecificationByPupilIdAndDates.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package school.journal.repository.specification.mark;
+
+import org.hibernate.criterion.Criterion;
+import org.hibernate.criterion.Restrictions;
+import school.journal.entity.Mark;
+import school.journal.repository.specification.Specification;
+
+import java.util.Date;
+
+public class MarkSpecificationByPupilIdAndDates extends MarkSpecification {
```
Comment (mbarouski) [inline server/src/main/java/school/journal/repository/specification/mark/MarkSpecificationByPupilIdAndDates.java]: delete dates

File: server/src/main/java/school/journal/service/IMarkService.java [hunk 10]
```diff
@@ -3,7 +3,13 @@
...
 
+import java.util.Date;
+import java.util.List;
+
 public interface IMarkService extends IService<Mark>{
 
-    public Mark getOne(int id) throws ServiceException ;
+    Mark getOne(int id) throws ServiceException ;
+    List<Mark> getMarksForSubjectInClass(int subjectId,int classId) throws ServiceException;
+    List<Mark> getMarksForTermOrder(int classId, Date startTerm,Date endTerm) throws ServiceException;
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/IMarkService.java]: delete dates

File: server/src/main/java/school/journal/service/impl/MarkService.java [hunk 11]
```diff
@@ -34,38 +39,38 @@ public MarkService(IRepository<Mark> repository) {
...
+            validateId(mark.getPupilId(), "Pupil");
+            validateId(mark.getSubjectId(), "Subject");
+            validateId(mark.getTeacherId(), "Teacher");
         } catch (ValidationException exc) {
             LOGGER.error(exc);
             throw new ServiceException(exc);
         }
         validateValue(mark.getValue());
-        validateDate(mark.getDate());
+        validateCreatingMarkDate(mark.getDate());
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/MarkService.java:46]: change name of method to more informative

File: server/src/main/java/school/journal/service/impl/MarkService.java [hunk 12]
```diff
@@ -34,38 +39,38 @@ public MarkService(IRepository<Mark> repository) {
...
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/MarkService.java:62]: we will ignore this date, so we can not check

File: server/src/main/java/school/journal/service/impl/MarkService.java [hunk 13]
```diff
@@ -106,7 +111,112 @@ public Mark getOne(int id) throws ServiceException {
...
+        Transaction transaction = session.beginTransaction();
+        //TODO Learn how to create complex criteries
+        //TODO DON'T DELETE THE CODE BELOW
+        /*Criteria criteria = session.createCriteria(Mark.class);
+        criteria.createAlias("school_journal_db.pupil", "pupil", INNER_JOIN).
+                add(Restrictions.and(
+                        new MarkSpecificationBySubjectId(subjectId).toCriteria(),
+                        Restrictions.eq("pupil.class_id", classId)
+                ));*/
+        List<Mark> markList;
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/MarkService.java]: test it

File: server/src/main/java/school/journal/service/impl/PupilService.java [hunk 14]
```diff
@@ -30,9 +29,9 @@ public PupilService(IRepository<Pupil> repository) {
     }
 
     @Override
-    public List<Pupil> getListOfPupils(int id) throws ServiceException {
+    public List<Pupil> getListOfPupils(int clazzId) throws ServiceException {
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/PupilService.java:33]: change name to `getListOfPupilsInClass`

File: server/src/main/java/school/journal/service/impl/SubjectInScheduleService.java [hunk 15]
```diff
@@ -21,15 +21,17 @@
 @Component("SubjectInScheduleService")
 public class SubjectInScheduleService extends CRUDService<SubjectInSchedule> implements ISubjectInScheduleService {
 
+    private static final long START_WORK_DAY_TIME_MILLIS = 25_200_000;//7h*60m*60s*1000ms
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/SubjectInScheduleService.java:24]: GOOD!

File: server/src/main/java/school/journal/utils/ValidateServiceUtils.java [hunk 16]
```diff
@@ -43,4 +45,11 @@ public static void validateEmail(String email) throws ValidationException {
             throw new ValidationException("Email isn't correct.");
         }
     }
+
+    public static void validateDatePeriod
+            (Date startDate, Date endDate) throws ValidationException{
```
Comment (mbarouski) [inline server/src/main/java/school/journal/utils/ValidateServiceUtils.java]: move to the new string only throwing

**Review by mbarouski — CHANGES_REQUESTED** [review]

**Comment by mbarouski** [comment]
and use some StringFormatter for sql query building

File: server/src/main/java/school/journal/controller/MarkAPIController.java (hunk 1 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/MarkAPIController.java]: done

File: server/src/main/java/school/journal/controller/MarkAPIController.java (hunk 2 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/MarkAPIController.java]: get mark of full time?

File: server/src/main/java/school/journal/controller/MarkAPIController.java (hunk 3 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/MarkAPIController.java]: done

File: server/src/main/java/school/journal/controller/MarkAPIController.java (hunk 4 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/MarkAPIController.java]: the same question

File: server/src/main/java/school/journal/controller/PupilAPIController.java (hunk 5 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/PupilAPIController.java:34]: What do you mean?

File: server/src/main/java/school/journal/controller/PupilAPIController.java (hunk 6 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/PupilAPIController.java]: done

File: server/src/main/java/school/journal/controller/PupilAPIController.java (hunk 7 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/PupilAPIController.java]: done

File: server/src/main/java/school/journal/controller/PupilAPIController.java (hunk 8 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/PupilAPIController.java]: What should i use?

File: server/src/main/java/school/journal/service/IMarkService.java (hunk 10 above)
Comment (zasamen) [reply server/src/main/java/school/journal/service/IMarkService.java]: id check it in COontroller

File: server/src/main/java/school/journal/service/impl/MarkService.java (hunk 12 above)
Comment (zasamen) [reply server/src/main/java/school/journal/service/impl/MarkService.java:62]: ?

File: server/src/main/java/school/journal/service/impl/PupilService.java (hunk 14 above)
Comment (zasamen) [reply server/src/main/java/school/journal/service/impl/PupilService.java:33]: done

File: server/src/main/java/school/journal/utils/ValidateServiceUtils.java (hunk 16 above)
Comment (zasamen) [reply server/src/main/java/school/journal/utils/ValidateServiceUtils.java]: done

File: server/src/main/java/school/journal/service/impl/MarkService.java (hunk 12 above)
Comment (zasamen) [reply server/src/main/java/school/journal/service/impl/MarkService.java:62]: deleted

File: server/src/main/java/school/journal/service/IMarkService.java (hunk 10 above)
Comment (zasamen) [reply server/src/main/java/school/journal/service/IMarkService.java]: ok

File: server/src/main/java/school/journal/controller/PupilAPIController.java (hunk 8 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/PupilAPIController.java]: What information should be in info like that?

File: server/src/main/java/school/journal/controller/PupilAPIController.java (hunk 5 above)
Comment (zasamen) [reply server/src/main/java/school/journal/controller/PupilAPIController.java:34]: ?

**Pull request merged**